packages/multichain-account-service/src/MultichainAccountWallet.ts

```diff
@@ -364,32 +364,28 @@ export class MultichainAccountWallet<
    * @returns The discovered accounts for each provider.
    */
   async discoverAndCreateAccounts(): Promise<Record<string, number>> {
-    const providers = this.#providers;
-    const providerContexts = new Map<
-      NamedAccountProvider,
-      {
-        stopped: boolean;
-        running?: Promise<void>;
-        groupIndex: number;
-        count: number;
-      }
-    >();
-
-    for (const p of providers) {
-      providerContexts.set(p, { stopped: false, groupIndex: 0, count: 0 });
-    }
-
-    let maxGroupIndex = 0;
-
-    const schedule = async (p: NamedAccountProvider, index: number) => {
-      // Ok to cast here because we know that the ctx will always be set.
-      const providerCtx = providerContexts.get(p) as {
-        stopped: boolean;
-        running?: Promise<void>;
-        groupIndex: number;
-        count: number;
-      };
-
+    // Start with the next available group index (so we can resume the discovery
+    // from there).
+    let maxGroupIndex = this.getNextGroupIndex();
+
+    const providerContexts: {
+      provider: NamedAccountProvider;
+      stopped: boolean;
+      running?: Promise<void>;
+      groupIndex: number;
+      count: number;
+    }[] = this.#providers.map((provider) => ({
+      provider,
+      stopped: false,
+      running: undefined,
+      groupIndex: maxGroupIndex,
+      count: 0,
+    }));
+
+    const schedule = async (
+      context: (typeof providerContexts)[number],
```

Comment:
Not sure about this though.. I thought of adding a local `type DiscoveryContext` (local to this function), WDYT?